docs/proposals/20230612-node-os-image-caching.md

```diff
@@ -0,0 +1,230 @@
+
+#### Credits
+
+A big thank you to Jack Francis, Michael Sinz, and Amr Hanafi for their work on the Kamino project that used a similar Prototype Pattern with AKSEngine which inspired this feature request: https://github.com/jackfrancis/kamino
+
+### Goals
+
+1. A solution using the Node Prototype pattern which cache’s nodes’ OS image on a regular interval and replaces it as a new ‘golden image’.
+1. Faster horizontal scale outs
+1. Prevent security breaking issues on node bootup from security updates being required immediately
+
+We will know we’ve succeeded when we can benchmark speed increases and successful image changes
+
+### Non-Goals/Future Work
+
+1. Extend the functionality to Windows nodes
+1. Optimization for efficiency and scalability
+1. A more complicated method of selecting a good candidate node such as incorporating manual forced prototype creation, which would perform better but take research to find the optimal method. 
+1. Optimization of the default time interval into a specific, best general default which would have to be researched
+
+## Proposal
+
+### User Stories
+
+#### As an operator of CAPZ, I want faster horizontal scaling on new node deployment
+
+People setting up CAPZ clusters will be able to have faster horizontal scaling because their Nodes’s OS images will be continually cached, and thus they will be able to avoid longer pull times and avoid having to create warm nodes and overprovision.
+
+#### As an operator, I want to be able to be more inline with security compliance as I spin up nodes
+
+As an operator I would like to be able to have my node’s OS image cache for the ability to avoid security alerts, cache security updates and patches so that I’ll be more security compliant and up to date when I have to spin up a new node. Otherwise, flags will be raised that a node is out of date on boot up before it finishes patching and a temporary security risk will exist.
+
+#### As an operator, I want to be able to have pods run quicker when scheduled
+
+As a cluster operator I want to bring up new nodes that already have the container images of the applications it will run pre-cached so that pods start running quicker when scheduled.
+
+### Requirements (Optional)
+
+#### Functional Requirements
+
+None currently known to be needed
+
+##### FR1
+
+##### FR2
+
+#### Non-Functional Requirements
+
+##### NFR1
+
+##### NFR2
+
+### Implementation Details/Notes/Constraints
+
+The plan is to modify the existing Controllers with the Node Prototype Pattern as desired. These controller additions can be added to AzureMachine Controller and AzureMachinePool Controller.
+
+An operator will be able to decide to turn the feature on or off with an environment variable on creation of a cluster, and then can update the AzureMachineTemplate and AzureMachinePool to customize how long they want the caching interval to be (see the yaml files below in this section for the caching interval).
```

Comment:
is the idea behind using an environment variable to turn this on via a feature flag for the whole CAPZ pod? Have we considered making it a more granular per cluster configuration so the feature can be enabled on a per-cluster basis?

Reply:
Made this much more understandable in the document - The environment variable will allow it to be usable for the whole CAPZ pod and then per cluster it will be default enabled or disabled (based on if we're in alpha, beta or graduated) and can be added or removed per cluster as desired by altering AzureMachineTemplate and AzureMachinePool instances.

Reply:
Since this absolutely should be a feature which is enable-able on a per-cluster basis